source/common/network/io_socket_handle_impl.cc

```diff
@@ -110,6 +127,7 @@ Api::IoCallUint64Result IoSocketHandleImpl::readv(uint64_t max_length, Buffer::R
 
 Api::IoCallUint64Result IoSocketHandleImpl::read(Buffer::Instance& buffer,
                                                  absl::optional<uint64_t> max_length_opt) {
+  // TODO(soulxu): there may have a more efficient way to read injected data instead of readv one.
```

Comment:
There also have recvmsg, recvmesg, recv to implement the reading from injected data, although there is no usecase for them yet.